src/cpu.c

```diff
@@ -23,9 +25,143 @@ void resetCpu(struct CPU * cpu, struct BUS * bus) {
   cpu->status.negative = false;
 }
 
+unsigned int absoluteAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned char lo = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned char hi = readBus(cpu->pc, bus);
+  cpu->pc++;
+  return (hi << 8) | lo;
+}
+
+unsigned int absoluteXAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned char lo = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned char hi = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned int address = (hi << 8) | lo;
+  return address + cpu->index_x;
+}
+
+unsigned int absoluteYAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned char lo = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned char hi = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned int address = (hi << 8) | lo;
+  return address + cpu->index_y;
+}
+
+unsigned int immediateAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned int address = cpu->pc;
+  cpu->pc++;
+  return address;
+}
+
+unsigned int indirectAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned int base_lo = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned int base_hi = readBus(cpu->pc, bus);
+  cpu->pc++;
+  unsigned int base = readBus((base_hi << 8) | base_lo, bus);
+  unsigned char lo = readBus(base, bus);
+  unsigned char hi = readBus(base+1, bus);
+  unsigned int address = (hi << 8) | lo;
+  return address;
+}
+
+unsigned int indirectXAddressMode(struct CPU *cpu, struct BUS *bus) {
+  unsigned int base = readBus(cpu->pc, bus) + cpu->index_x;
```

Comment:
Same as previous comment, can we encapsulate indirect functions?

Reply:
Indirect have some different behaviors between them, I extracted some redundant code anyways.